blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,866 @@
+    local_addrs: Vec<Address>,
+    pending: HashMap<Address, MsgSet>,
+    pub(crate) cur_tipset: Tipset,
+    api: T,
+    pub min_gas_price: BigInt,
+    pub max_tx_pool_size: i64,
+    pub network_name: String,
+    bls_sig_cache: LruCache<Cid, Signature>,
+    sig_val_cache: LruCache<String, ()>,
+    local_msgs: HashMap<Vec<u8>, SignedMessage>,
+}
+
+impl<T> MessagePool<T>
+where
+    T: Provider,
+{
+    /// Create a new message pool
+    pub fn new(mut api: T, network_name: String) -> Result<MessagePool<T>, Error>
+    where
+        T: Provider,
+    {
+        let tipset = api.subscribe_head_changes();
+        // LruCache sizes have been taken from the lotus implementation
+        let bls_sig_cache = LruCache::new(40000);
+        let sig_val_cache = LruCache::new(32000);
+        let mut mp = MessagePool {
+            local_addrs: Vec::new(),
+            pending: HashMap::new(),
+            cur_tipset: tipset,
+            api,
+            min_gas_price: ToBigInt::to_bigint(&0).unwrap(),
+            max_tx_pool_size: 5000,
+            network_name,
+            bls_sig_cache,
+            sig_val_cache,
+            local_msgs: HashMap::new(),
+        };
+        mp.load_local()?;
+        Ok(mp)
+    }
+
+    /// Add a signed message to local_addrs and local_msgs
+    fn add_local(&mut self, m: &SignedMessage, msgb: Vec<u8>) -> Result<(), Error> {
+        self.local_addrs.push(*m.from());
+        self.local_msgs.insert(msgb, m.clone());
+        Ok(())
+    }
+
+    /// Push a signed message to the MessagePool
+    pub fn push(&mut self, msg: &SignedMessage) -> Result<Cid, Error> {
+        let msg_serial = msg
+            .marshal_cbor()
+            .map_err(|err| Error::Other(err.to_string()))?;
+        self.add(msg)?;
+        self.add_local(msg, msg_serial)?;
+        msg.cid().map_err(|err| Error::Other(err.to_string()))
+    }
+
+    /// This is a helper to push that will help to make sure that the message fits the parameters
+    /// to be pushed to the MessagePool
+    pub fn add(&mut self, msg: &SignedMessage) -> Result<(), Error> {
+        let size = msg
+            .marshal_cbor()
+            .map_err(|err| Error::Other(err.to_string()))?
+            .len();
+        if size > 32 * 1024 {
+            return Err(Error::MessageTooBig);
+        }
+        if msg
+            .value()
+            .gt(&ToBigUint::to_biguint(&2_000_000_000).unwrap())
+        {
+            return Err(Error::MessageValueTooHigh);
+        }
+
+        self.verify_msg_sig(msg)?;
+
+        let tmp = msg.clone();
+        self.add_tipset(tmp, &self.cur_tipset.clone())
+    }
+
+    /// Add a SignedMessage without doing any of the checks
+    pub fn add_skip_checks(&mut self, m: SignedMessage) -> Result<(), Error> {
+        self.add_helper(m)
+    }
+
+    /// Return the string representation of the message signature
+    pub(crate) fn sig_cache_key(&mut self, msg: &SignedMessage) -> Result<String, Error> {
```

Comment:
This shouldn't be returning a string, the only reason go impl uses string is because it's an immutable version of a byte slice